controller/src/main/java/io/pravega/controller/store/kvtable/InMemoryKVTMetadataStore.java

```diff
@@ -98,20 +97,25 @@ public Scope newScope(String scopeName) {
     @Override
     @Synchronized
     public CompletableFuture<Boolean> checkScopeExists(String scope) {
-        return CompletableFuture.completedFuture(scopes.containsKey(scope));
+        return Futures.completeOn(CompletableFuture.completedFuture(this.streamStore.scopeExists(scope)), executor);
     }
 
     @Override
     @Synchronized
     public CompletableFuture<Boolean> checkTableExists(String scopeName, String kvt) {
-        return CompletableFuture.completedFuture((InMemoryScope) getScope(scopeName)).thenApply(scope -> scope.checkTableExists(kvt));
+        return Futures.completeOn(checkScopeExists(scopeName).thenCompose(exists -> {
+            if (exists) {
+                return CompletableFuture.completedFuture(((InMemoryScope) getScope(scopeName)).checkTableExists(kvt));
+            }
+            return CompletableFuture.completedFuture(Boolean.FALSE);
+        }), executor);
     }
 
     @Override
     @Synchronized
     public Scope getScope(final String scopeName) {
-        if (scopes.containsKey(scopeName)) {
-            return scopes.get(scopeName);
+        if (this.streamStore.scopeExists(scopeName)) {
+            return this.streamStore.getScope(scopeName);
         } else {
             return new InMemoryScope(scopeName);
```

Comment:
is there a check that ensures that "create" is called on the scope object before it is used. 
this is returning the scope object, but where does it get added to the store?

Reply:
This is implemented in same lines as similar method in InMemoryMetadataStore:
https://github.com/pravega/pravega/blob/master/controller/src/main/java/io/pravega/controller/store/stream/InMemoryStreamMetadataStore.java#L117

Reply:
yes, but it looks incorrect in both places. we need to see if getScope should return scope not found if the scope does not exist in the store.

Reply:
I agree, but it may need significant code refactoring as there are multiple dependent methods that would break.
Filed: https://github.com/pravega/pravega/issues/5021